tools/infrasys/src/main.rs

```diff
@@ -0,0 +1,872 @@
+                    missing: format!("pub_key_threshold for '{}' repo config", repo_name),
+                })?,
+            &root_role_path.display().to_string(),
+        )?;
+        sign_root(
+            root_keys,
+            &KeyRole::Root,
+            repo_config
+                .root_key_threshold
+                .as_ref()
+                .context(error::MissingConfig {
+                    missing: format!("root_key_threshold for '{}' repo config", repo_name),
+                })?,
+            &root_role_path.display().to_string(),
+        )?;
+
+        // Step 5: Upload root.json
+        info!("Uploading root.json to S3 bucket...");
+        upload_file(s3_region, &bucket_name, prefix, root_role_path).await?;
+
+        // Step 6: Update output paramters if not already set
+        if repo_config.metadata_base_url.is_none() {
+            repo_config.metadata_base_url = Some(
+                Url::parse(format!("{}{}/metadata/", &bucket_url, prefix).as_str())
+                    .context(error::ParseUrl { input: &bucket_url })?,
+            );
+        }
+        if repo_config.targets_url.is_none() {
+            repo_config.targets_url = Some(
+                Url::parse(format!("{}{}/targets/", &bucket_url, prefix).as_str())
+                    .context(error::ParseUrl { input: &bucket_url })?,
+            );
+        }
+        if repo_config.root_role_url.is_none() {
+            repo_config.root_role_url = Some(
+                Url::parse(format!("{}{}/root.json", &bucket_url, prefix).as_str())
+                    .context(error::ParseUrl { input: &bucket_url })?,
+            );
+
+            let root_role_data = fs::read_to_string(&root_role_path).context(error::FileRead {
+                path: root_role_path,
+            })?;
+            let mut d = Sha512::new();
+            d.update(&root_role_data);
+            let digest = hex::encode(d.finalize());
+            repo_config.root_role_sha512 = Some(digest);
+        }
+    }
+
+    //Step 7: Generate Infra.lock
+    info!("Writing Infra.lock...");
+    let yaml_string = serde_yaml::to_string(&infra_config).context(error::InvalidYaml)?;
+    fs::write(
+        toml_path
+            .parent()
+            .context(error::Parent { path: toml_path })?
+            .join("Infra.lock"),
+        yaml_string,
+    )
+    .context(error::FileWrite { path: toml_path })?;
+
+    info!("Complete!");
+    Ok(())
+}
+
+//  =^..^=   =^..^=   =^..^= S3 BUCKET FUNCTIONALITY  =^..^=   =^..^=   =^..^=
+
+/// Creates a *private* S3 Bucket using a CloudFormation template
+/// Input: The region in which the bucket will be created and the name of the bucket
+/// Output: The stack_arn of the bucket created (will be added as a field to Infra.lock)
+async fn create_s3_bucket(
+    region: &String,
+    stack_name: &String,
+) -> Result<(String, String, String)> {
+    // IN-FUTURE: Add support for accomodating pre-existing buckets (skip this creation process)
```

Comment:
I think it's ok to change all `IN-FUTURE` tags to `TODO` tags so it's easier for other people to find and reference. The `TODO` don't have to mean that you have to finish them in the PR, but just indicates future work.